metrics/cgroups/metrics.go

```diff
@@ -59,9 +60,10 @@ func taskID(id, namespace string) string {
 type collector struct {
 	mu sync.RWMutex
 
-	cgroups map[string]*task
-	ns      *metrics.Namespace
-	metrics []*metric
+	cgroups      map[string]*task
+	ns           *metrics.Namespace
+	metrics      []*metric
+	finalMetrics map[string]([]prometheus.Metric)
```

Comment:
I would almost make this a channel `storedMetrics: make(chan prometheus.Metric, someSize)` and then remove the last metric method and just call collect with `c.storedMetrics`.

Maybe make the buffer size N metrics * 100 or something to keep up to 100 container metrics in memory.  Just incase nothing reads from containerd, we don't want to fill up the memory with metrics.  That or we can run a flush goroutine that will force flush the queue every so often to make sure we don't block or use too much memory.

Reply:
Wouldn't a buffered channel stall a number of goroutines that would otherwise fill up the buffer, leading to a filled up memory in the same way? Or do I change the send loop in metric.go to a nonblocking one?

Reply:
Ok, I looked at the prom code.  I it has a channel buffer size of 1000 so changing `metric.go` to be nonblocking could mean we miss some metrics from prom when it is collecting.  I'm still thinking of a good way to handle these two caes.

Reply:
On the other hand, the prom code starts draining the channel right after this, so it should not be too much of an issue unless the system is under fairly heavy load (although 1000 is not that much, since there's around 60 metrics for each container due to containerd/metrics/cgroups alone).